File: valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []

        opens = {"(": ")", "{": "}", "[": "]"}
        closes = [")", "}", "]"]

        # Loop over each bracket in the string
        for brace in s:

            # Check if the bracket is a type of open bracket.
            if brace in opens:
                
                # Append the closing bracket
                stack.append(opens[brace])

            # If the bracket is a type of closing bracket
            elif brace in closes:
                
                # Check if the top of the stack matches the type of bracket
                if not len(stack) or brace != stack.pop():
                    return False


        # Check if the list is empty at the end.
        return len(stack) == 0

File: valid-parentheses/valid-parentheses.cs
private bool IsValid(string expression)
{
	Stack<char> stack = new();
	Dictionary<char, char> validMatches = new() { { '(', ')' }, { '{', '}' }, { '[', ']' } };
	List<char> closingBrackets = new() { ')', '}', ']'};

	// Loop over each bracket in the string
	foreach (var brace in expression)
	{
		// Check if the bracket is a type of open bracket.
		if (validMatches.ContainsKey(brace))
		{
			// Append the closing bracket
			stack.Push(validMatches[brace]);
		}
		// If the bracket is a type of closing bracket
		else if (closingBrackets.Contains(brace))
		{
			// Check if the top of the stack matches the type of bracket
			if (stack.Count == 0 || brace != stack.Pop())
			{
				return false;
			}
		}
	}

	// Check if the list is empty at the end.
	return stack.Count == 0;
}